File: beginner/pagina13/3231/3231.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int main() {
    int n, h, l;
    cin >> n >> h >> l;

    vector<int> horror_list(h);
    for (int i = 0; i < h; i++)
        cin >> horror_list[i];

    vector<vector<int>> graph(n);
    const int INF = 1e9;
    vector<int> dist(n, INF);
    for (int i = 0; i < l; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    queue<int> fila;
    for (int x : horror_list) {
        dist[x] = 0;
        fila.push(x);
    }

    while (!fila.empty()) {
        int curr = fila.front();
        fila.pop();
        for (int v : graph[curr]) {
            if (dist[v] == INF) {
                dist[v] = dist[curr] + 1;
                fila.push(v);
            }
        }
    }

    int maxH = dist[0];
    for (int i = 1; i < n; i++)
        maxH = max(maxH, dist[i]);

    for (int i = 0; i < n; i++) {
        if (dist[i] == maxH) {
            cout << i << '\n';
            break;
        }
    }
    return 0;
}

File: beginner/pagina13/3231/3231.py
"""
    HI:

    se o filme(id) estiver na lista -> Hi == 0
    se não tem similaridade com nenhum filme da lista -> HI = INFINITO
    do contrário -> HI = distância que o filme tem de um filme da lista.
    
"""

import sys
input = sys.stdin.buffer.readline

INFINIT = float('inf')
n, h, l = map(int, input().split())
horror_list = map(int, input().split())
graph = {id: [] for id in range(n) }

for _ in range(l):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

queue = []
dist = [INFINIT for _ in range(n)]
for id in horror_list:
    queue.append(id)
    dist[id] = 0

head = 0
while head < len(queue):
    curr = queue[head]
    head += 1
    for v in graph[curr]:
        if dist[v] == INFINIT:
            dist[v] = dist[curr] + 1
            queue.append(v)

maxH = max(dist)

for id in range(n):
    if dist[id] == maxH:
        print(id)
        break